### PR TITLE
SEAL 3.6

### PR DESCRIPTION
**Pull request by bcebere**
## Description
SEAL 3.6

Changelog: https://github.com/microsoft/SEAL/releases/tag/v3.6.0

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -6,6 +6,7 @@ set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_STANDARD_REQUIRED True)
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR})
+include_directories(third_party/SEAL/native/src)
 
 # ##############################################################################
 # Third party
```
Comment (youben11) [inline CMakeLists.txt:12]: We can remove the comments here as we no longer build it using VS, right?

File: tenseal/cpp/context/sealcontext.cpp [hunk 2]
```diff
@@ -28,16 +28,15 @@ EncryptionParameters create_bfv_parameters(size_t poly_modulus_degree,
...
         CoeffModulus::Create(poly_modulus_degree, coeff_mod_bit_sizes));
 
     return parameters;
 }
 
-shared_ptr<seal::SEALContext> create_context(EncryptionParameters parms) {
-    auto context = SEALContext::Create(parms);
-    return context;
+SEALContext create_context(EncryptionParameters parms) {
+    return SEALContext(parms);
```
Comment (youben11) [inline tenseal/cpp/context/sealcontext.cpp:40]: Is there any reason for this change?

File: tenseal/cpp/context/tensealencoder.h [hunk 3]
```diff
@@ -115,7 +115,7 @@ class TenSEALEncoder {
     /*
     Stores a shared_pointer to the SEAL Context.
     */
-    std::shared_ptr<SEALContext> _context;
+    SEALContext _context;
```
Comment (youben11) [inline tenseal/cpp/context/tensealencoder.h:118]: So we now keep a full copy of the context instead of a pointer? Why is that?

**Review by youben11 — CHANGES_REQUESTED** [review]
Looks great! I left some questions here that we can discuss later over the meeting

File: tenseal/cpp/context/sealcontext.cpp (hunk 2 above)
Comment (bcebere) [reply tenseal/cpp/context/sealcontext.cpp:40]: SEALContext::Create was removed in this release

File: tenseal/cpp/context/tensealencoder.h (hunk 3 above)
Comment (bcebere) [reply tenseal/cpp/context/tensealencoder.h:118]: They dropped the Create call, which created the shared_ptr, and they are copying the SealContexts everytime.

Actually, I kept the shared_ptr in the TenSEAL context, we can do that here as well.

File: tenseal/cpp/context/sealcontext.cpp (hunk 2 above)
Comment (bcebere) [reply tenseal/cpp/context/sealcontext.cpp:40]: This is their commit
https://github.com/microsoft/SEAL/commit/a94374d348fb28390e6191cbed52830d67617736#diff-797d0099b848a1dbe9380766c6034f9f500c85b708ec931ea6dc7aa9520f3767

File: tenseal/cpp/context/tensealencoder.h (hunk 3 above)
Comment (youben11) [reply tenseal/cpp/context/tensealencoder.h:118]: Yeah, I guess it doesn't require a full copy, as we will be mainly accessing attributes, or do you have a reason to believe we shouldn't use a pointer here?

File: tenseal/cpp/context/sealcontext.cpp (hunk 2 above)
Comment (youben11) [reply tenseal/cpp/context/sealcontext.cpp:40]: Thanks! I couldn't find anything about it in the release note, as I was wondering if it's only a design choice, or that there is implications in having multiple holders of the same context.

File: tenseal/cpp/context/tensealencoder.h (hunk 3 above)
Comment (bcebere) [reply tenseal/cpp/context/tensealencoder.h:118]: I will revert it here. Initially, I was going to drop std::shared_ptr<SEALContext> everywhere, and this is some partial change.

File: tenseal/cpp/context/sealcontext.cpp (hunk 2 above)
Comment (bcebere) [reply tenseal/cpp/context/sealcontext.cpp:40]: While is not great to change the API that much, they do not need a shared ptr, their context is pretty light.

**Review by youben11 — APPROVED** [review]

**Pull request merged**